### ISSUE TITLE
Create Variable and Scope classes

### ISSUE THREAD

**Issue by lucaswoj** [manifold-lang/manifold#25]
A `Scope` object` holds all variables and values in a particular function or file or namespace.

I already have this mostly done.

The interface is something like

``` java
public class Scope {
  Scope(Scope parentScope);

  public void defineVariable(Expression typeExpression, VariableIdentifier identifier);
  public TypeValue getVariableTypeValue(VariableIdentifier identifier);
  public Variable getVariable(VariableIdentifier identifier);
  public Value getVariableValue(VariableIdentifier identifier);
  public void assignVariable(VariableIdentifier identifier, Expression valueExpression);

  public Scope getParentScope();
}
```

``` java
public class Variable {
  public Variable(TypeValue typeValue, VariableIdentifier identifier);
  public boolean isAssigned();
  public Value getValue();
  public void setValue(Value value);
}
```

**Comment by mtrberzi** [comment]
I have some groundwork for this done in MurphEngineer/whdl/master but it still needs exceptions and test cases (and doesn't do much yet).

**Comment by mtrberzi** [comment]
Pull request outstanding for this issue.

**Comment by lucaswoj** [comment]
Closed by #38
